### DISCUSSION TITLE
Incorrect Order of Merged Documents in Merge Operation

### DISCUSSION THREAD

**Discussion by vivek781113** [gotenberg/gotenberg#674] [Q&A]
The merge operation within the service is not merging documents in the correct order, resulting in the output merged document having an unexpected sequence. 

Below is the snap of postman request. The merged document is having the sequence as 3 then 1 then 2. Any help would be appreciated.
![image](https://github.com/gotenberg/gotenberg/assets/32304458/1c825ebc-d930-44a8-b5bd-8d7a1233ab13)

Gotenberg service 
![image](https://github.com/gotenberg/gotenberg/assets/32304458/f87601e7-9da5-42f8-bab8-3ef27dd48aa7)

**Comment by gulien** [comment]
Hello @vivek781113,

That’s weird. Could you share the debug logs?

**Reply by vivek781113** [reply]
Hi Gulien, please find below the, 

Curl request:
`curl --location 'http://localhost:3000/forms/pdfengines/merge' \
--form 'chin.pdf=@"/C:/Users/vivekkumar.tiwary/Desktop/test files/chinise.pdf"' \
--form 'hin.pdf=@"/C:/Users/vivekkumar.tiwary/Desktop/test files/Hindi.pdf"' \
--form 'korean=@"/C:/Users/vivekkumar.tiwary/Desktop/test files/korean-lang.pdf"'`

Logs:
[_pdfservice_logs.txt](https://github.com/gotenberg/gotenberg/files/12393706/_pdfservice_logs.txt)


**_Current output: Hindi.pdf --> chinise.pdf --> korean.pdf
Expected output: chinese.pdf --> Hindi.pdf --> Korean.pdf_**

**Reply by gulien** [reply]
Using version `7.9.1` on the demo instance, it worked as expected:

```
curl --location 'https://demo.gotenberg.dev/forms/pdfengines/merge' \
--form 'chin.pdf=@"/chin.pdf"' \
--form 'hin.pdf=@"/hin.pdf"' \
--form 'korean.pdf=@"/korean.pdf"'
```

Got  the correct order, i.e., chin.pdf then hin.pdf then korean.pdf.

**Reply by jimmy-zhuENV** [reply]
Has anyone recently seen this problem again and found a solution? I am seeing this incorrect ordering on the latest version 8 and in version 7 (including 7.9.1).

**Reply by gulien** [reply]
@jimmy-zhuENV what’s your exact problem?

**Reply by jimmy-zhuENV** [reply]
same problem as the one described above. if i run this in the terminal
curl --location 'https://demo.gotenberg.dev/forms/pdfengines/merge' \
--form 'chin.pdf=@"/chin.pdf"' \
--form 'hin.pdf=@"/hin.pdf"' \
--form 'korean.pdf=@"/korean.pdf"'
-o ~/Downloads/merged.pdf

The merged pdf is in the wrong order. (eg chin.pdf, korean.pdf, hin.pdf)

I have tried multiple versions of gotenberg locally and always see this same problem.

**Reply by gulien** [reply]
Just tested it on the demo instance, and it's working perfectly fine 🤔 

[PDF_Engines_Merge-1709630433515.pdf](https://github.com/gotenberg/gotenberg/files/14493093/PDF_Engines_Merge-1709630433515.pdf)

Are there some hidden characters in your filename?

**Reply by jimmy-zhuENV** [reply]
These were the files that I was trying to merge in order.
Shouldnt be any hidden characters as I created these from a blank word document.

[Part_ONE.pdf](https://github.com/gotenberg/gotenberg/files/14501792/Part_ONE.pdf)
[Part_TWO.pdf](https://github.com/gotenberg/gotenberg/files/14501793/Part_TWO.pdf)
[Part_THREE.pdf](https://github.com/gotenberg/gotenberg/files/14501794/Part_THREE.pdf)

**Reply by gulien** [reply]
In this case, they will be merged in the following order:

1. Part_ONE.pdf
2. Part_THREE.pdf
3. Part_TWO.pdf

(Alphabetically, ONE > THREE > TWO).

With Gotenberg 8.2, if you don't want to play with letters, just prefix your files with numbers to set the merge order.

**Reply by jimmy-zhuENV** [reply]
This makes a lot of sense. Thanks for the clarification!
